qiita_db/support_files/patches/33.sql

```diff
@@ -0,0 +1,435 @@
+
+-- Publication table - holds the minimum information for a given publication
+-- It is useful to keep track of the publication of the studies and the software
+-- used for processing artifacts
+CREATE TABLE qiita.publication (
+    doi                  varchar  NOT NULL,
+    pubmed_id            integer  ,
+    CONSTRAINT pk_publication PRIMARY KEY ( doi )
+ ) ;
+
+-- Software publictation table - relates each software package with the lists of
+-- its related publciations
+CREATE TABLE qiita.software_publication (
+    software_id          bigint  NOT NULL,
+    publication_doi      varchar  NOT NULL,
+    CONSTRAINT idx_software_publication_0 PRIMARY KEY ( software_id, publication_doi )
+ ) ;
+CREATE INDEX idx_software_publication_software ON qiita.software_publication ( software_id ) ;
+CREATE INDEX idx_software_publication_publication ON qiita.software_publication ( publication_doi ) ;
+ALTER TABLE qiita.software_publication ADD CONSTRAINT fk_software_publication FOREIGN KEY ( software_id ) REFERENCES qiita.software( software_id )    ;
+ALTER TABLE qiita.software_publication ADD CONSTRAINT fk_software_publication_0 FOREIGN KEY ( publication_doi ) REFERENCES qiita.publication( doi )    ;
+
+-- Add remaining FK to the artifact table. Creating here since the target tables
+-- do not exist when the artifact table was created
+ALTER TABLE qiita.artifact ADD CONSTRAINT fk_artifact_visibility FOREIGN KEY ( visibility_id ) REFERENCES qiita.visibility( visibility_id )    ;
+ALTER TABLE qiita.artifact ADD CONSTRAINT fk_artifact_soft_command FOREIGN KEY ( command_id ) REFERENCES qiita.soft_command( command_id )    ;
+
+-- Create a function to infer the visibility of the artifact from the
+-- raw data
+CREATE FUNCTION infer_rd_status(rd_id bigint, st_id bigint) RETURNS bigint AS $$
+    DECLARE
+        result bigint;
+    BEGIN
+        CREATE TEMP TABLE irds_temp
+            ON COMMIT DROP AS
+                SELECT DISTINCT processed_data_status_id
+                    FROM qiita.processed_data
+                        JOIN qiita.preprocessed_processed_data USING (processed_data_id)
+                        JOIN qiita.prep_template_preprocessed_data USING (preprocessed_data_id)
+                        JOIN qiita.prep_template USING (prep_template_id)
+                        JOIN qiita.study_processed_data USING (processed_data_id)
+                    WHERE raw_data_id = rd_id AND study_id = st_id;
+        IF EXISTS(SELECT * FROM irds_temp WHERE processed_data_status_id = 2) THEN
+            result := 2;
+        ELSIF EXISTS(SELECT * FROM irds_temp WHERE processed_data_status_id = 3) THEN
+            result := 3;
+        ELSIF EXISTS(SELECT * FROM irds_temp WHERE processed_data_status_id = 1) THEN
+            result := 1;
+        ELSE
+            result := 4;
+        END IF;
+        DROP TABLE irds_temp;
```

Comment:
Your solution is not correct - the order of the if statements is important. I first need to check for 2, then for 3, then for 1 and if none of them are present, then 4. That's why I can't use a min and I have to use an exist to check each of them independently. I can use multiple queries, but since it is a huge joint, I think is better to retrieve the data once and just query it.